File: examples/web_runtime/museum-client.js
/**
 * Museum Client - Web Runtime
 * Browser-based museum tour player with offline support
 * Mobile-optimized with audio, QR scanning, and analytics
 */

class MuseumClient {
    constructor() {
        this.story = null;
        this.currentPassage = null;
        this.visited = {};
        this.variables = {};
        this.sessionStart = Date.now();
        this.audioPlayedCount = 0;
        this.language = 'en';

        // Audio player state
        this.audioElement = null;
        this.audioPlaying = false;

        // Initialize after DOM is ready
        if (document.readyState === 'loading') {
            document.addEventListener('DOMContentLoaded', () => this.initialize());
        } else {
            this.initialize();
        }
    }

    initialize() {
        // Get audio element
        this.audioElement = document.getElementById('audio-element');

        // Set up event listeners
        this.setupEventListeners();
        this.setupAudioListeners();

        console.log('✅ Museum Client initialized');
    }

    /* ==========================================
       STORY MANAGEMENT
       ========================================== */

    loadStory(story) {
        this.story = story;

        // Initialize variables
        if (story.variables) {
            story.variables.forEach(variable => {
                this.variables[variable.name] = variable.initial;
            });
        }

        // Set UI metadata
        const metadata = story.metadata || {};
        document.getElementById('tour-title').textContent = metadata.title || 'Museum Tour';
        document.getElementById('tour-subtitle').textContent = metadata.museum || '';

        console.log('✅ Story loaded:', metadata.title);
    }

    start() {
        if (!this.story) {
            console.error('No story loaded');
            return;
        }

        // Find start passage
        const startId = this.story.settings?.startPassage || this.story.passages[0]?.id || 'welcome';
        this.gotoPassage(startId);
    }

    findPassage(identifier) {
        if (!this.story) return null;

        return this.story.passages.find(p =>
            p.id === identifier || p.name === identifier
        );
    }

    /* ==========================================
       NAVIGATION
       ========================================== */

    gotoPassage(identifier) {
        const passage = this.findPassage(identifier);

        if (!passage) {
            console.error('Passage not found:', identifier);
            return false;
        }

        this.currentPassage = passage;

        // Track visit
        if (!this.visited[passage.id]) {
            this.visited[passage.id] = {
                firstVisit: Date.now(),
                visitCount: 0
            };
        }

        this.visited[passage.id].visitCount++;
        this.visited[passage.id].lastVisit = Date.now();

        // Update variables
        if (this.variables.visited_count !== undefined) {
            this.variables.visited_count = this.getVisitedCount();
        }

        // Render passage
        this.render();
        this.updateProgress();

        // Scroll to top
        document.querySelector('.museum-main').scrollTo(0, 0);

        return true;
    }

    scanQR(qrCode) {
        // Find passage with matching QR code
        const passage = this.story.passages.find(p =>
            p.metadata && p.metadata.qrCode === qrCode
        );

        if (passage) {
            this.gotoPassage(passage.id);
            this.closeModal('qr-modal');
            return true;
        }

        alert('No exhibit found for QR code: ' + qrCode);
        return false;
    }

    /* ==========================================
       RENDERING
       ========================================== */

    render() {
        if (!this.currentPassage) return;

        const passage = this.currentPassage;

        // Title
        document.getElementById('passage-title').textContent = passage.name || passage.id;

        // Metadata
        this.renderMetadata(passage.metadata);

        // Content
        document.getElementById('passage-content').innerHTML = this.processContent(passage.text);

        // Choices
        this.renderChoices(passage.choices || []);
    }

    renderMetadata(metadata) {
        const container = document.getElementById('passage-metadata');
        container.innerHTML = '';

        if (!metadata) return;

        const badges = [];

        if (metadata.floor) {
            badges.push({ icon: '📍', text: `Floor ${metadata.floor}` });
        }

        if (metadata.hasAudio && metadata.audioLength) {
            badges.push({ icon: '🎧', text: metadata.audioLength });
        }

        if (metadata.qrCode) {
            badges.push({ icon: '🔲', text: metadata.qrCode });
        }

        if (metadata.popularity) {
            const stars = '⭐'.repeat(metadata.popularity);
            badges.push({ icon: '', text: stars });
        }

        badges.forEach(badge => {
            const el = document.createElement('div');
            el.className = 'metadata-badge';
            el.innerHTML = `
                ${badge.icon ? `<span class="icon">${badge.icon}</span>` : ''}
                <span>${badge.text}</span>
            `;
            container.appendChild(el);
        });
    }

    processContent(text) {
        if (!text) return '';

        // Process variables {{variable}}
        text = text.replace(/\{\{([a-zA-Z_][a-zA-Z0-9_]*)\}\}/g, (match, varName) => {
            if (this.variables.hasOwnProperty(varName)) {
                return String(this.variables[varName]);
            }
            return match;
        });

        // Process markdown
        // Headers
        text = text.replace(/^# (.+)$/gm, '<h2>$1</h2>');
        text = text.replace(/^## (.+)$/gm, '<h3>$1</h3>');
        text = text.replace(/^### (.+)$/gm, '<h4>$1</h4>');

        // Bold and italic
        text = text.replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>');
        text = text.replace(/\*(.+?)\*/g, '<em>$1</em>');

        // Audio/image references
        text = text.replace(/\[audio\]\(([^)]+)\)/g, (match, path) => {
            return `<button class="inline-audio-btn" onclick="museumClient.playAudioFromPath('${path}')" aria-label="Play audio">🎧 Play Audio Guide</button>`;
        });

        text = text.replace(/\[image\]\(([^)]+)\)/g, (match, path) => {
            return `<img src="${path}" alt="Exhibit image" loading="lazy">`;
        });

        // Lists (simple implementation)
        text = text.replace(/^- (.+)$/gm, '<li>$1</li>');
        text = text.replace(/(<li>.*<\/li>\n?)+/g, '<ul>$&</ul>');

        // Paragraphs
        const paragraphs = text.split(/\n\n+/).filter(p => p.trim());
        text = paragraphs.map(p => {
            // Don't wrap if already has block-level tags
            if (p.match(/^<(h[1-6]|ul|ol|div|button)/)) {
                return p;
            }
            return `<p>${p.trim()}</p>`;
        }).join('');

        return text;
    }

    renderChoices(choices) {
        const container = document.getElementById('passage-choices');
        container.innerHTML = '';

        choices.forEach((choice, index) => {
            const btn = document.createElement('button');
            btn.className = 'choice-btn';
            btn.textContent = choice.text;
            btn.setAttribute('aria-label', `Choice: ${choice.text}`);

            btn.addEventListener('click', () => {
                this.gotoPassage(choice.target);
            });

            container.appendChild(btn);
        });
    }

    /* ==========================================
       PROGRESS TRACKING
       ========================================== */

    updateProgress() {
        const visitedCount = this.getVisitedCount();
        const totalCount = this.story.passages.length;
        const percentage = Math.floor((visitedCount / totalCount) * 100);

        document.getElementById('progress-bar').style.width = percentage + '%';
        document.getElementById('progress-text').textContent = `${visitedCount}/${totalCount} exhibits`;
    }

    getVisitedCount() {
        return Object.keys(this.visited).length;
    }

    /* ==========================================
       AUDIO PLAYER
       ========================================== */

    playAudioFromPath(audioPath) {
        const fullPath = '../museum_tours/natural_history/' + audioPath;
        this.playAudio(fullPath, 'Audio Guide');
    }

    playAudio(audioPath, title = 'Audio Guide') {
        if (!this.audioElement) return;

        this.audioElement.src = audioPath;
        this.audioElement.load();
        this.audioElement.play();

        // Show audio player
        const player = document.getElementById('audio-player');
        player.style.display = 'block';
        setTimeout(() => player.classList.add('visible'), 10);

        // Set title
        document.getElementById('audio-title').textContent = title;

        // Track play count
        this.audioPlayedCount++;

        if (this.variables.audio_count !== undefined) {
            this.variables.audio_count = this.audioPlayedCount;
        }
    }

    toggleAudioPlayPause() {
        if (!this.audioElement) return;

        if (this.audioElement.paused) {
            this.audioElement.play();
        } else {
            this.audioElement.pause();
        }
    }

    seekAudio(delta) {
        if (!this.audioElement) return;
        this.audioElement.currentTime = Math.max(0, this.audioElement.currentTime + delta);
    }

    closeAudioPlayer() {
        const player = document.getElementById('audio-player');
        player.classList.remove('visible');
        setTimeout(() => {
            player.style.display = 'none';
            if (this.audioElement) {
                this.audioElement.pause();
                this.audioElement.src = '';
            }
        }, 300);
    }

    setupAudioListeners() {
        if (!this.audioElement) return;

        // Play/pause button
        const playIcon = document.getElementById('audio-play-icon');

        this.audioElement.addEventListener('play', () => {
            playIcon.textContent = '⏸️';
        });

        this.audioElement.addEventListener('pause', () => {
            playIcon.textContent = '▶️';
        });

        // Update duration and seek bar
        this.audioElement.addEventListener('loadedmetadata', () => {
            this.updateAudioDuration();
        });

        this.audioElement.addEventListener('timeupdate', () => {
            this.updateAudioDuration();
            this.updateAudioSeekBar();
        });

        // Seek bar input
        const seekBar = document.getElementById('audio-seek-bar');
        seekBar.addEventListener('input', (e) => {
            const time = (e.target.value / 100) * this.audioElement.duration;
            this.audioElement.currentTime = time;
        });

        // Control buttons
        document.getElementById('audio-play-pause').addEventListener('click', () => {
            this.toggleAudioPlayPause();
        });

        document.getElementById('audio-rewind').addEventListener('click', () => {
            this.seekAudio(-15);
        });

        document.getElementById('audio-forward').addEventListener('click', () => {
            this.seekAudio(15);
        });

        document.getElementById('audio-close').addEventListener('click', () => {
            this.closeAudioPlayer();
        });
    }

    updateAudioDuration() {
        if (!this.audioElement) return;

        const current = this.formatTime(this.audioElement.currentTime);
        const duration = this.formatTime(this.audioElement.duration);

        document.getElementById('audio-duration').textContent = `${current} / ${duration}`;
    }

    updateAudioSeekBar() {
        if (!this.audioElement) return;

        const percentage = (this.audioElement.currentTime / this.audioElement.duration) * 100;
        document.getElementById('audio-seek-bar').value = percentage || 0;
    }

    formatTime(seconds) {
        if (isNaN(seconds)) return '0:00';

        const mins = Math.floor(seconds / 60);
        const secs = Math.floor(seconds % 60);
        return `${mins}:${secs.toString().padStart(2, '0')}`;
    }

    /* ==========================================
       MAP
       ========================================== */

    showMap() {
        const container = document.getElementById('map-container');
        container.innerHTML = '';

        // Create exhibit list
        const listEl = document.createElement('div');
        listEl.className = 'map-exhibit-list';

        // Get exhibits (passages with exhibitId metadata)
        const exhibits = this.story.passages.filter(p => p.metadata && p.metadata.exhibitId);

        // Sort by floor
        exhibits.sort((a, b) => {
            const floorA = a.metadata.floor || 0;
            const floorB = b.metadata.floor || 0;
            return floorA - floorB;
        });

        exhibits.forEach(exhibit => {
            const visited = this.visited[exhibit.id];
            const exhibitEl = document.createElement('div');
            exhibitEl.className = 'map-exhibit' + (visited ? ' visited' : '');

            exhibitEl.innerHTML = `
                <span class="exhibit-status">${visited ? '✓' : '◯'}</span>
                <div class="exhibit-info">
                    <div class="exhibit-name">${exhibit.name}</div>
                    <div class="exhibit-details">
                        Floor ${exhibit.metadata.floor || '?'} •
                        ${exhibit.metadata.hasAudio ? '🎧 Audio' : 'No audio'} •
                        QR: ${exhibit.metadata.qrCode || 'N/A'}
                    </div>
                </div>
            `;

            exhibitEl.addEventListener('click', () => {
                this.gotoPassage(exhibit.id);
                this.closeModal('map-modal');
            });

            listEl.appendChild(exhibitEl);
        });

        container.appendChild(listEl);
        this.showModal('map-modal');
    }

    /* ==========================================
       STATISTICS
       ========================================== */

    showStats() {
        const container = document.getElementById('stats-container');
        container.innerHTML = '';

        // Calculate stats
        const visitedCount = this.getVisitedCount();
        const totalCount = this.story.passages.length;
        const percentage = Math.floor((visitedCount / totalCount) * 100);
        const duration = Math.floor((Date.now() - this.sessionStart) / 60000);

        // Create stats grid
        const gridEl = document.createElement('div');
        gridEl.className = 'stats-grid';

        const stats = [
            { value: visitedCount, label: 'Exhibits Visited' },
            { value: `${percentage}%`, label: 'Completion' },
            { value: `${duration} min`, label: 'Tour Duration' },
            { value: this.audioPlayedCount, label: 'Audio Guides' }
        ];

        stats.forEach(stat => {
            const cardEl = document.createElement('div');
            cardEl.className = 'stat-card';
            cardEl.innerHTML = `
                <div class="stat-value">${stat.value}</div>
                <div class="stat-label">${stat.label}</div>
            `;
            gridEl.appendChild(cardEl);
        });

        container.appendChild(gridEl);

        // Most visited exhibit
        if (visitedCount > 0) {
            let mostVisited = null;
            let maxVisits = 0;

            for (const [passageId, info] of Object.entries(this.visited)) {
                if (info.visitCount > maxVisits) {
                    maxVisits = info.visitCount;
                    mostVisited = this.findPassage(passageId);
                }
            }

            if (mostVisited) {
                const detailEl = document.createElement('div');
                detailEl.className = 'help-section';
                detailEl.innerHTML = `
                    <h3>⭐ Most Revisited</h3>
                    <p><strong>${mostVisited.name}</strong> (${maxVisits} times)</p>
                `;
                container.appendChild(detailEl);
            }
        }

        this.showModal('stats-modal');
    }

    /* ==========================================
       SESSION EXPORT
       ========================================== */

    exportSession() {
        const sessionData = {
            storyIfid: this.story.metadata?.ifid,
            sessionStart: this.sessionStart,
            durationSeconds: Math.floor((Date.now() - this.sessionStart) / 1000),
            visited: this.visited,
            variables: this.variables,
            audioPlayedCount: this.audioPlayedCount,
            language: this.language
        };

        // Download as JSON
        const blob = new Blob([JSON.stringify(sessionData, null, 2)], { type: 'application/json' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = `museum-visit-${Date.now()}.json`;
        a.click();
        URL.revokeObjectURL(url);

        alert('Visit data exported! Check your downloads.');
    }

    /* ==========================================
       MODALS
       ========================================== */

    showModal(modalId) {
        const modal = document.getElementById(modalId);
        if (modal) {
            modal.classList.add('visible');

            // Close on overlay click
            const overlay = modal.querySelector('.modal-overlay');
            overlay.addEventListener('click', () => this.closeModal(modalId));
        }
    }

    closeModal(modalId) {
        const modal = document.getElementById(modalId);
        if (modal) {
            modal.classList.remove('visible');
        }
    }

    /* ==========================================
       EVENT LISTENERS
       ========================================== */

    setupEventListeners() {
        // Navigation buttons
        document.getElementById('btn-menu').addEventListener('click', () => this.showModal('menu-modal'));
        document.getElementById('btn-map').addEventListener('click', () => this.showMap());
        document.getElementById('btn-qr').addEventListener('click', () => this.showModal('qr-modal'));
        document.getElementById('btn-audio').addEventListener('click', () => {
            if (this.currentPassage?.metadata?.hasAudio) {
                // Play current passage audio if available
                alert('Audio feature coming soon!');
            } else {
                alert('No audio available for this exhibit');
            }
        });
        document.getElementById('btn-stats').addEventListener('click', () => this.showStats());
        document.getElementById('btn-help').addEventListener('click', () => this.showModal('help-modal'));

        // Modal close buttons
        document.getElementById('map-modal-close').addEventListener('click', () => this.closeModal('map-modal'));
        document.getElementById('qr-modal-close').addEventListener('click', () => this.closeModal('qr-modal'));
        document.getElementById('stats-modal-close').addEventListener('click', () => this.closeModal('stats-modal'));
        document.getElementById('help-modal-close').addEventListener('click', () => this.closeModal('help-modal'));
        document.getElementById('menu-modal-close').addEventListener('click', () => this.closeModal('menu-modal'));

        // QR Scanner
        document.getElementById('qr-submit').addEventListener('click', () => {
            const qrCode = document.getElementById('qr-input').value.trim();
            if (qrCode) {
                this.scanQR(qrCode);
            }
        });

        document.getElementById('qr-input').addEventListener('keypress', (e) => {
            if (e.key === 'Enter') {
                document.getElementById('qr-submit').click();
            }
        });

        // Menu options
        document.getElementById('menu-restart').addEventListener('click', () => {
            if (confirm('Are you sure you want to restart the tour?')) {
                this.visited = {};
                this.variables = {};
                this.sessionStart = Date.now();
                this.audioPlayedCount = 0;
                this.start();
                this.closeModal('menu-modal');
            }
        });

        document.getElementById('menu-export').addEventListener('click', () => {
            this.exportSession();
            this.closeModal('menu-modal');
        });

        document.getElementById('menu-about').addEventListener('click', () => {
            alert('Museum Tour powered by Whisker\nVersion 1.0.0\n\nCreated with Claude Code');
        });

        // Keyboard shortcuts
        document.addEventListener('keydown', (e) => {
            if (e.key === 'Escape') {
                // Close any open modals
                document.querySelectorAll('.museum-modal.visible').forEach(modal => {
                    modal.classList.remove('visible');
                });

                // Close audio player
                if (document.getElementById('audio-player').classList.contains('visible')) {
                    this.closeAudioPlayer();
                }
            }
        });
    }
}

// Global instance
const museumClient = new MuseumClient();

File: examples/museum_tours/museum_client.lua
#!/usr/bin/env lua
-- Museum Tour Client (CLI)
-- Cross-platform service-oriented client for running Whisker museum tours
-- Foundation for iOS/Android native app implementations

local json = require("src.utils.json")

-- Museum Client Class
local MuseumClient = {}
MuseumClient.__index = MuseumClient

-- Create new museum client instance
function MuseumClient.new()
    local self = setmetatable({}, MuseumClient)
    self.story = nil
    self.current_passage = nil
    self.visited = {}
    self.variables = {}
    self.session_start = os.time()
    self.audio_played_count = 0
    self.language = "en"
    return self
end

--------------------------------------------------------------------------------
-- STORY LOADING
--------------------------------------------------------------------------------

-- Load a museum tour story from file
function MuseumClient:load_story(filepath)
    local file = io.open(filepath, "r")
    if not file then
        return nil, "Could not open story file: " .. filepath
    end

    local content = file:read("*all")
    file:close()

    local story, err = json.decode(content)
    if not story then
        return nil, "Failed to parse JSON: " .. tostring(err)
    end

    -- Validate story structure
    if not story.passages or #story.passages == 0 then
        return nil, "Story has no passages"
    end

    self.story = story

    -- Initialize variables
    if story.variables then
        for _, var in ipairs(story.variables) do
            self.variables[var.name] = var.initial
        end
    end

    -- Find start passage
    local start_id = "welcome"
    if story.settings and story.settings.startPassage then
        start_id = story.settings.startPassage
    end

    self.current_passage = self:find_passage(start_id)
    if not self.current_passage then
        return nil, "Could not find start passage: " .. start_id
    end

    return true, nil
end

-- Find a passage by ID or name
function MuseumClient:find_passage(identifier)
    if not self.story then
        return nil
    end

    for _, passage in ipairs(self.story.passages) do
        if passage.id == identifier or passage.name == identifier then
            return passage
        end
    end

    return nil
end

--------------------------------------------------------------------------------
-- NAVIGATION
--------------------------------------------------------------------------------

-- Navigate to a passage
function MuseumClient:goto_passage(identifier)
    local passage = self:find_passage(identifier)
    if not passage then
        return false, "Passage not found: " .. identifier
    end

    self.current_passage = passage

    -- Track visit
    if not self.visited[passage.id] then
        self.visited[passage.id] = {
            first_visit = os.time(),
            visit_count = 0
        }
    end

    self.visited[passage.id].visit_count = self.visited[passage.id].visit_count + 1
    self.visited[passage.id].last_visit = os.time()

    -- Update variables
    if self.variables.visited_count then
        self.variables.visited_count = self:get_visited_count()
    end

    return true, nil
end

-- Scan QR code (simulate by entering exhibit ID)
function MuseumClient:scan_qr(qr_code)
    -- Find passage with matching QR code in metadata
    for _, passage in ipairs(self.story.passages) do
        if passage.metadata and passage.metadata.qrCode == qr_code then
            return self:goto_passage(passage.id)
        end
    end

    return false, "No exhibit found for QR code: " .. qr_code
end

-- Get available choices for current passage
function MuseumClient:get_choices()
    if not self.current_passage then
        return {}
    end

    return self.current_passage.choices or {}
end

-- Choose an option
function MuseumClient:choose(choice_index)
    local choices = self:get_choices()

    if choice_index < 1 or choice_index > #choices then
        return false, "Invalid choice: " .. choice_index
    end

    local choice = choices[choice_index]
    return self:goto_passage(choice.target)
end

--------------------------------------------------------------------------------
-- DISPLAY
--------------------------------------------------------------------------------

-- Render current passage to terminal
function MuseumClient:render()
    if not self.current_passage then
        print("No passage loaded")
        return
    end

    -- Clear screen (cross-platform)
    os.execute("clear") or os.execute("cls")

    -- Header
    self:print_header()

    -- Passage content
    print("\n" .. string.rep("=", 80))
    self:print_content(self.current_passage.text)
    print(string.rep("=", 80))

    -- Metadata (audio, QR code, etc.)
    if self.current_passage.metadata then
        self:print_metadata(self.current_passage.metadata)
    end

    -- Choices
    local choices = self:get_choices()
    if #choices > 0 then
        print("\n📍 Choose your next destination:")
        for i, choice in ipairs(choices) do
            print(string.format("  [%d] %s", i, choice.text))
        end
    end

    -- Commands
    print("\n" .. string.rep("-", 80))
    print("Commands: [number] Navigate | [m]ap | [q]r scan | [s]tats | [h]elp | [x] Exit")
    print(string.rep("-", 80))
end

-- Print header with tour info
function MuseumClient:print_header()
    if not self.story or not self.story.metadata then
        return
    end

    local meta = self.story.metadata
    print("\n🏛️  " .. (meta.title or "Museum Tour"))

    if meta.museum then
        print("📍 " .. meta.museum)
    end

    if meta.estimatedTime then
        print("⏱️  Estimated time: " .. meta.estimatedTime)
    end

    -- Progress
    local visited_count = self:get_visited_count()
    local total = #self.story.passages
    local percent = math.floor((visited_count / total) * 100)
    print(string.format("📊 Progress: %d/%d exhibits (%d%%)", visited_count, total, percent))
end

-- Print formatted content (handle markdown-style formatting)
function MuseumClient:print_content(text)
    if not text then
        return
    end

    -- Simple markdown-style rendering for terminal
    local lines = {}
    for line in text:gmatch("[^\n]+") do
        -- Headers
        if line:match("^# ") then
            line = "\n" .. line:gsub("^# ", ""):upper()
        elseif line:match("^## ") then
            line = "\n" .. line:gsub("^## ", "")
        elseif line:match("^### ") then
            line = "  " .. line:gsub("^### ", "")
        end

        -- Bullet points
        if line:match("^%- ") then
            line = "  • " .. line:gsub("^%- ", "")
        end

        -- Audio/image references
        line = line:gsub("%[audio%]%((.-)%)", "🎧 Audio available: %1")
        line = line:gsub("%[image%]%((.-)%)", "🖼️  Image: %1")

        table.insert(lines, line)
    end

    print(table.concat(lines, "\n"))
end

-- Print passage metadata
function MuseumClient:print_metadata(metadata)
    local info = {}

    if metadata.hasAudio and metadata.audioLength then
        table.insert(info, "🎧 Audio guide: " .. metadata.audioLength)
    end

    if metadata.floor then
        table.insert(info, "📍 Floor " .. metadata.floor)
    end

    if metadata.qrCode then
        table.insert(info, "🔲 QR Code: " .. metadata.qrCode)
    end

    if metadata.popularity then
        local stars = string.rep("⭐", metadata.popularity)
        table.insert(info, "Popularity: " .. stars)
    end

    if #info > 0 then
        print("\n" .. table.concat(info, " | "))
    end
end

--------------------------------------------------------------------------------
-- FEATURES
--------------------------------------------------------------------------------

-- Show museum map (list all exhibits)
function MuseumClient:show_map()
    os.execute("clear") or os.execute("cls")
    print("\n🗺️  MUSEUM MAP\n")
    print(string.rep("=", 80))

    local exhibits = {}
    for _, passage in ipairs(self.story.passages) do
        if passage.metadata and passage.metadata.exhibitId then
            local status = self.visited[passage.id] and "✓" or " "
            local floor = passage.metadata.floor or "?"
            local audio = passage.metadata.hasAudio and "🎧" or "  "

            table.insert(exhibits, {
                name = passage.name,
                floor = floor,
                audio = audio,
                status = status,
                qr = passage.metadata.qrCode
            })
        end
    end

    -- Sort by floor
    table.sort(exhibits, function(a, b)
        if a.floor == b.floor then
            return a.name < b.name
        end
        return a.floor < b.floor
    end)

    -- Display
    for _, exhibit in ipairs(exhibits) do
        print(string.format("[%s] Floor %s %s  %s  (QR: %s)",
            exhibit.status, exhibit.floor, exhibit.audio, exhibit.name, exhibit.qr or "N/A"))
    end

    print(string.rep("=", 80))
    print("\nPress Enter to continue...")
    io.read()
end

-- Show visitor statistics
function MuseumClient:show_stats()
    os.execute("clear") or os.execute("cls")
    print("\n📊 VISIT STATISTICS\n")
    print(string.rep("=", 80))

    -- Duration
    local duration_seconds = os.time() - self.session_start
    local minutes = math.floor(duration_seconds / 60)
    print(string.format("⏱️  Tour duration: %d minutes", minutes))

    -- Exhibits visited
    local visited_count = self:get_visited_count()
    local total = #self.story.passages
    local percent = math.floor((visited_count / total) * 100)
    print(string.format("📍 Exhibits visited: %d/%d (%d%%)", visited_count, total, percent))

    -- Audio guides
    print(string.format("🎧 Audio guides played: %d", self.audio_played_count))

    -- Most popular passage
    local most_visited = nil
    local max_visits = 0
    for passage_id, info in pairs(self.visited) do
        if info.visit_count > max_visits then
            max_visits = info.visit_count
            most_visited = self:find_passage(passage_id)
        end
    end

    if most_visited then
        print(string.format("⭐ Most revisited: %s (%d times)", most_visited.name, max_visits))
    end

    -- Variables
    if self.variables and next(self.variables) then
        print("\n📝 Tour Variables:")
        for name, value in pairs(self.variables) do
            print(string.format("  • %s = %s", name, tostring(value)))
        end
    end

    print(string.rep("=", 80))
    print("\nPress Enter to continue...")
    io.read()
end

-- Play audio (simulated - would trigger native audio player on mobile)
function MuseumClient:play_audio(audio_path)
    print(string.format("\n🎧 Playing audio: %s", audio_path))
    print("   (Audio playback would be handled by native player on mobile)")
    self.audio_played_count = self.audio_played_count + 1

    -- Update variables
    if self.variables.audio_count then
        self.variables.audio_count = self.audio_played_count
    end
end

-- Get count of unique passages visited
function MuseumClient:get_visited_count()
    local count = 0
    for _ in pairs(self.visited) do
        count = count + 1
    end
    return count
end

-- Export session data (for analytics or resume)
function MuseumClient:export_session()
    return {
        story_ifid = self.story.metadata and self.story.metadata.ifid,
        session_start = self.session_start,
        duration_seconds = os.time() - self.session_start,
        visited = self.visited,
        variables = self.variables,
        audio_played_count = self.audio_played_count,
        language = self.language
    }
end

--------------------------------------------------------------------------------
-- COMMAND LOOP
--------------------------------------------------------------------------------

-- Show help
function MuseumClient:show_help()
    os.execute("clear") or os.execute("cls")
    print([[

🏛️  MUSEUM TOUR CLIENT - HELP

NAVIGATION:
  [1-9]     Choose a numbered option to navigate to that exhibit
  [m]ap     View museum floor plan and all exhibits
  [q]r      Scan a QR code (enter exhibit QR code)

INFORMATION:
  [s]tats   View your visit statistics
  [h]elp    Show this help screen

ACTIONS:
  [x]       Exit tour and save progress

TIPS:
  • Visit exhibits in any order you prefer
  • Use QR codes on exhibit labels for quick access
  • Check the map to see your progress
  • Audio guides enhance your experience

Press Enter to continue...
]])
    io.read()
end

-- Main command loop
function MuseumClient:run()
    if not self.story then
        print("Error: No story loaded")
        return
    end

    print("\n🏛️  Welcome to the Museum Tour Client")
    print("Type 'h' for help, 'x' to exit\n")
    print("Press Enter to begin...")
    io.read()

    while true do
        self:render()

        -- Get user input
        io.write("\n> ")
        local input = io.read()

        if not input or input == "x" or input == "X" then
            -- Exit
            self:show_stats()
            print("\n👋 Thank you for visiting! Your progress has been saved.\n")
            break

        elseif input == "h" or input == "H" then
            -- Help
            self:show_help()

        elseif input == "m" or input == "M" then
            -- Map
            self:show_map()

        elseif input == "s" or input == "S" then
            -- Stats
            self:show_stats()

        elseif input == "q" or input == "Q" then
            -- QR scan
            print("Enter QR code (e.g., MUSEUM-DINO-001):")
            io.write("> ")
            local qr_code = io.read()
            local success, err = self:scan_qr(qr_code)
            if not success then
                print("❌ " .. err)
                print("Press Enter to continue...")
                io.read()
            end

        elseif tonumber(input) then
            -- Numeric choice
            local choice_num = tonumber(input)
            local success, err = self:choose(choice_num)
            if not success then
                print("❌ " .. err)
                print("Press Enter to continue...")
                io.read()
            end

        else
            print("❌ Unknown command. Type 'h' for help.")
            print("Press Enter to continue...")
            io.read()
        end
    end
end

--------------------------------------------------------------------------------
-- MAIN ENTRY POINT
--------------------------------------------------------------------------------

local function main(args)
    if #args < 1 then
        print([[
Usage: lua museum_client.lua <story_file.whisker>

Example:
  lua museum_client.lua natural_history/story.whisker

This is a cross-platform CLI client for running Whisker museum tours.
It serves as a reference implementation and foundation for native mobile apps.
]])
        return 1
    end

    local story_path = args[1]

    -- Create client
    local client = MuseumClient.new()

    -- Load story
    local success, err = client:load_story(story_path)
    if not success then
        print("Error loading story: " .. err)
        return 1
    end

    -- Run tour
    client:run()

    return 0
end

-- Run if executed directly
if arg and arg[0]:match("museum_client%.lua$") then
    os.exit(main(arg))
end

-- Export for use as module
return MuseumClient
